### PR TITLE
Add wait_for_rollout to kubernetes_stateful_set

### LINKED ISSUE #605: Wait for StatefulSet rollout status
Hi,

Like **Deployment Resource**, **StatefulSet** rollout status should be successful before kubernetes provider decides that it is successfully deployed. Is there any obstacles no to use same kind of strategy used in Deployment Resource(**waitForDeploymentReplicasFunc**)? If not, I would gladly to contribute with a similar implementation.

### Terraform&Providers' Version
- Terraform v0.12.7
- terraform-providers/kubernetes: 1.9.0

### Affected Resource(s)
Please list the resources as a list, for example:
- **StatefulSet**

### Expected Behavior
Provider should wait before **StatefulSet** is deployed&running state.

### Actual Behavior
Provider successfully terminates saying that **StatefulSet** is created successfully.

### Steps to Reproduce
Please list the steps required to reproduce the issue;
1. `terraform apply` with a StatefulSet resource.

### Important Factoids
Running at GKE

### PR DESCRIPTION
**Pull request by jrhouston**
### Description

This PR adds a `wait_for_rollout` attribute to the `kubernetes_stateful_set` resource. 

A gotcha I discovered with this resource is that it doesn't set any conditions like Deployment and Job do, so I had to find a different way of determining if the rollout was done. You also can't use `kubectl wait` with this resource. There is an open issue about it here: https://github.com/kubernetes/kubernetes/issues/79606

I discovered that kubectl exposes a helper package with a`StatusViewer` object that implements the logic for determining if a StatefulSet is done. Annoyingly, this package only works with unstructured types, so I had to convert the response from the API before supplying it to this package. I still think this is better than duplicating the rollout logic from kubectl in the provider.

There is a bunch of vendor spam because I had to bring in the `k8s.io/kubectl` package.

Resolves #605 

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)

### References

https://github.com/kubernetes/kubernetes/issues/79606
https://godoc.org/k8s.io/kubectl/pkg/polymorphichelpers

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Comment by jrhouston** [comment]
- [ ] Need to do this for the Update operation too

File: kubernetes/resource_kubernetes_stateful_set.go [hunk 1]
```diff
@@ -203,3 +227,46 @@ func resourceKubernetesStatefulSetDelete(d *schema.ResourceData, meta interface{
 
 	return nil
 }
+
+// retryUntilStatefulSetRolloutComplete checks if a give job finished its execution and either in Complete or Failed state
```
Comment (dak1n1) [inline kubernetes/resource_kubernetes_stateful_set.go:231]:
```suggestion
// retryUntilStatefulSetRolloutComplete checks if a given job finished its execution and is either in 'Complete' or 'Failed' state.
```

File: kubernetes/resource_kubernetes_stateful_set.go [hunk 2]
```diff
@@ -203,3 +227,46 @@ func resourceKubernetesStatefulSetDelete(d *schema.ResourceData, meta interface{
...
+func retryUntilStatefulSetRolloutComplete(conn *kubernetes.Clientset, ns, name string) resource.RetryFunc {
+	return func() *resource.RetryError {
+		res, err := conn.AppsV1().StatefulSets(ns).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		// NOTE This is what kubectl uses to determine if a rollout is done
```
Comment (dak1n1) [inline kubernetes/resource_kubernetes_stateful_set.go:239]:
```suggestion
		// NOTE: This is what kubectl uses to determine if a rollout is done.
```

File: kubernetes/resource_kubernetes_stateful_set.go [hunk 3]
```diff
@@ -203,3 +227,46 @@ func resourceKubernetesStatefulSetDelete(d *schema.ResourceData, meta interface{
...
+		// We are using this here because the logic for determining if a StatefulSet
+		// is done is gnarly and we dont want to duplicate it in the provider
```
Comment (dak1n1) [inline kubernetes/resource_kubernetes_stateful_set.go:241]: Ha, I like the personality this comment has. Just adding punctuation.

```suggestion
		// is done is gnarly and we don't want to duplicate it in the provider.
```

File: kubernetes/resource_kubernetes_stateful_set.go [hunk 4]
```diff
@@ -203,3 +227,46 @@ func resourceKubernetesStatefulSetDelete(d *schema.ResourceData, meta interface{
...
+		gvk := appsv1.SchemeGroupVersion.WithKind("StatefulSet")
+		gk := gvk.GroupKind()
+		statusViewer, err := polymorphichelpers.StatusViewerFor(gk)
```
Comment (dak1n1) [inline kubernetes/resource_kubernetes_stateful_set.go:244]: Nice, it's great there was a function for this.

File: kubernetes/resource_kubernetes_stateful_set.go [hunk 5]
```diff
@@ -203,3 +227,46 @@ func resourceKubernetesStatefulSetDelete(d *schema.ResourceData, meta interface{
...
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(res)
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		// NOTE For some reason the kind and apiVersion get lost when converting to unstructured
```
Comment (dak1n1) [inline kubernetes/resource_kubernetes_stateful_set.go:254]:
```suggestion
		// NOTE: For some reason, the Kind and apiVersion get lost when converting to unstructured.
```

File: kubernetes/resource_kubernetes_stateful_set.go [hunk 6]
```diff
@@ -203,3 +227,46 @@ func resourceKubernetesStatefulSetDelete(d *schema.ResourceData, meta interface{
...
+		obj["apiVersion"] = gvk.GroupVersion().String()
+		obj["kind"] = gvk.Kind
+		u := unstructured.Unstructured{Object: obj}
+
+		// NOTE the revision parameter of the Status function below is not actually used
```
Comment (dak1n1) [inline kubernetes/resource_kubernetes_stateful_set.go:259]:
```suggestion
		// NOTE: the revision parameter of the Status function below is not actually used.
```

File: kubernetes/resource_kubernetes_stateful_set_test.go [hunk 7]
```diff
@@ -365,6 +365,7 @@ func testAccCheckKubernetesStatefulSetExists(n string, obj *api.StatefulSet) res
 
 func testAccKubernetesStatefulSetChecksBasic(name string) resource.TestCheckFunc {
 	return resource.ComposeAggregateTestCheckFunc(
+		resource.TestCheckResourceAttr(statefulSetTestResourceName, "wait_for_rollout", "true"),
```
Comment (dak1n1) [inline kubernetes/resource_kubernetes_stateful_set_test.go:368]: Cool, this actually helps teach me what I need to do today for my tests :)

**Comment by jrhouston** [comment]
Thanks @dak1n1 - I had to rework the test for this to actually check that it had rolled out or not, see my latest commit.

**Pull request merged**